### PR TITLE
Return a 400 response to HTTP requests

### PR DESCRIPTION
**Pull request by davel**
For our deployment, it is useful if coturn returns a valid HTTP response to an HTTP request. To do this on the same port as STUN/TURN and without enabling the admin site, I have extended `read_client_connection()` to return a canned HTTP response, in response to an HTTP request, rather than immediately closing the connection.

### CODE REVIEW

**Comment by eakraly** [comment]
Thank you @davel !

Since this is a very specific use case we should have a flag to enable this functionality (and have it disabled by default).
This will preserve the existing behavior for everyone today and allow you to have your specific behavior with a flag.

Does it make sense?

**Comment by davel** [comment]
Yes, I can add a flag for this, I shall take a look at the weekend.
 
The change *should* be safe to enable in general. It would only break a workflow which expects an empty response to an HTTP request.

**Comment by davel** [comment]
> Thank you @davel !
> 
> Since this is a very specific use case so it should have a flag to enable this functionality (and have it disabled by default). This will preserve the existing behavior for everyone today and allow you to have your specific behavior with a flag.
> 
> Does it make sense?

This is now done, there is now a `unsupported-http-gives-400` parameter.

File: src/server/ns_turn_server.c [hunk 1]
```diff
@@ -4535,6 +4535,29 @@ static int read_client_connection(turn_turnserver *server, ts_ur_super_session *
...
+          } else {
+            TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "%s: %s returning 400 response", __FUNCTION__, proto);
+          }
+
+          ioa_network_buffer_handle nbh_http = ioa_network_buffer_allocate(ss->client_socket->e);
+
+          char *content = "HTTP Not supported.\n";
+          char buffer[1024];
+          snprintf(buffer, sizeof(buffer),
+                   "HTTP/1.1 400 %s Not supported\r\nConnection: close\r\nServer: %s\r\nContent-Type: "
```
Comment (eakraly) [inline src/server/ns_turn_server.c:4555]: Let's keep http headers before the content, not after

**Review by eakraly — CHANGES_REQUESTED** [review]
I have reservations regarding the setting name: let's find a more generic name
Maybe `respond-http-unsupported`?

**Comment by davel** [comment]
> I have reservations regarding the setting name: let's find a more generic name Maybe `respond-http-unsupported`?

I have change the option (and variable names) to --respond-http-unsupported.

File: src/server/ns_turn_server.c (hunk 1 above)
Comment (davel) [reply src/server/ns_turn_server.c:4555]: > Let's keep http headers before the content, not after

That would be neater, however we need to know the length of the content before we can generate the header.

File: src/server/ns_turn_server.c [hunk 2]
```diff
@@ -4535,6 +4535,29 @@ static int read_client_connection(turn_turnserver *server, ts_ur_super_session *
...
+                   "text/plain\r\nContent-Length: %d\r\n\r\n%s",
+                   proto, TURN_SOFTWARE, (int)strlen(content), content);
```
Comment (ggarber) [inline src/server/ns_turn_server.c:4557]: Are you sure you want to disclose the TURN_SOFTWARE here?    I would remove it for any server in production so probably it is better to never include it in this response.

**Comment by ggarber** [comment]
The change looks ok to me.  I would review the issue with disclosing the server name and version for security reasons given that this is in the public interface.

**Review by ggarber — APPROVED** [review]

**Comment by davel** [comment]
I added the software version for consistency with the the other http listeners in coturn.  It is not necessary for my application.

**Comment by davel** [comment]
I am happy for the version and server name to be removed from the response. I don't have hardware with me to do so myself for a few days.

File: src/server/ns_turn_server.c (hunk 1 above)
Comment (eakraly) [reply src/server/ns_turn_server.c:4555]: You can generated the body, measure its length, and then construct the response

Comment (davel) [reply src/server/ns_turn_server.c]: > You can generated the body, measure its length, and then construct the response

@eakraly  I have added comments and moved the length into a variable to make it clear that it is doing this in that order.

**Comment by davel** [comment]
@eakraly I believe I have resolved all the issues you raised and CI is passing again since merging in current `master`. Is there anything else I need to do to this PR for it to be accepted?

**Review by eakraly — APPROVED** [review]

**Pull request merged**